src/game/level/parser.rs

```diff
@@ -1,315 +1,283 @@
+}
+
+impl From<LexError> for LevelParsingError {
+	fn from(value: LexError) -> Self {
+		Self::LexError(value)
 	}
+}
 
-	if !cycle_layout.iter().all(Option::is_some) {
-		return Err(LevelParsingError::MissingCycleLayout);
+impl From<LevelBuilderError> for LevelParsingError {
+	fn from(value: LevelBuilderError) -> Self {
+		Self::BuilderError(value)
 	}
+}
 
-	let mut layout: Vec<DeclaredPlacement> = cycle_layout
-		.into_iter()
-		.enumerate()
-		.map(|(i, x)| {
-			let (placement, hints) = x.unwrap();
-			DeclaredPlacement::Cycle(DeclaredCyclePlacement {
-				cycle_index: i,
-				position: placement.position,
-				radius: placement.radius,
-				double_intersection_hints: hints,
-			})
-		})
-		.collect();
-
-	for (line_id, statement) in statements.iter() {
-		if let Statement::PlaceVertex(data) = statement {
-			if data.len() < 2 {
-				return Err(LevelParsingError::MalformedStatement(*line_id));
+impl std::error::Error for LevelParsingError {}
+
+impl std::fmt::Display for LevelParsingError {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			LevelParsingError::InvalidKeyword(k) => write!(f, "The keyword {} is invalid.", k)?,
+			LevelParsingError::InvalidModifier(m) => write!(f, "The modifier [{}] is invalid.", m)?,
+			LevelParsingError::UnknownVertexName(name) => write!(
+				f,
+				"A vertex with name {} was referenced, but not defined.",
+				name
+			)?,
+			LevelParsingError::UnknownCycleName(name) => write!(
+				f,
+				"A cycle with name {} was referenced, but not defined.",
+				name
+			)?,
+			LevelParsingError::RedefinedCycle(name) => {
+				write!(f, "A cycle with name {} was defined multiple times.", name)?
+			}
+			LevelParsingError::LexError(e) => e.fmt(f)?,
+			LevelParsingError::MissingModifier => write!(f, "Statement is missing a modifier.")?,
+			LevelParsingError::ArgumentIsNotANumber(arg) => {
+				write!(f, "Expected a numeric argument, got {arg}.")?
 			}
-			let vertex_name = data[0];
-			let angle_str = data[1].replace(',', ".");
-			if let Some(vertex_id) = vertex_names.get(vertex_name) {
-				let Ok(angle) = angle_str.parse::<f32>() else {
-					return Err(LevelParsingError::MalformedStatement(*line_id));
-				};
-				layout.push(DeclaredPlacement::Vertex(DeclaredVertexPlacement {
-					vertex_index: *vertex_id,
-					relative_angle: angle,
-				}))
-			} else {
-				return Err(LevelParsingError::UnknownCycleName(vertex_name.to_string()));
+			LevelParsingError::InvalidMetaVariable(name) => {
+				write!(f, "{name} is not a valid meta variable name")?
 			}
+			LevelParsingError::BuilderError(e) => e.fmt(f)?,
+			LevelParsingError::NotEnoughArguments(needed, got) => write!(
+				f,
+				"Statement found {got} arguments, needs at least {needed}."
+			)?,
 		}
+		Ok(())
 	}
-
-	Ok(LevelFile {
-		metadata,
-		data: LevelData {
-			vertices,
-			cycles,
-			linkages,
-		},
-		layout,
-	})
 }
 
-#[test]
-fn basic_test() {
-	let data = r"
-TEST6=?!#_AAA 648
```

Comment:
note: This was present to make sure the parser doesn't crap itself on an unknown assignment key (it should spit out a warning or something probably.)